### PR TITLE
Add back build breaking functionality on test/work item failure

### PR DESCRIPTION
**Pull request by jonfortescue**
cc / @MattGal @markwilkie @chcosta

### CODE REVIEW

File: src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs [hunk 1]
```diff
@@ -48,6 +48,22 @@ private async Task WaitForHelixJobAsync(string jobName)
...
+                    if ((fail ?? 0) > 0)
+                    {
+                        Log.LogError($"{fail ?? -1} work item(s) have failed.");
+                    }
+
+                    int? testFailures;
+                    workItemSummary[0].Data.Analysis[0].Status.TryGetValue("fail", out testFailures);
+                    if ((testFailures ?? 0) > 0)
+                    {
+                        Log.LogError($"{testFailures ?? -1} tests have failed.");
```
Comment (MattGal) [inline src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs]: Nit - there's no benefit to coalescing null to -1 given that you can't get into this branch with a null testFailures.

Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs]: true. I guess if it does fail catastrophically we'd rather see a stack trace here anyway.

Comment (alexperovich) [reply src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs]: You should include that job that had the failure in this message.

Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs]: Ah, good idea.

File: src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs [hunk 2]
```diff
@@ -48,6 +48,22 @@ private async Task WaitForHelixJobAsync(string jobName)
...
+                    int? fail;
+                    workItemSummary[0].Data.WorkItemStatus.TryGetValue("fail", out fail);
+
+                    if ((fail ?? 0) > 0)
+                    {
+                        Log.LogError($"{fail} work item(s) have failed.");
+                    }
+
+                    int? testFailures;
+                    workItemSummary[0].Data.Analysis[0].Status.TryGetValue("fail", out testFailures);
```
Comment (MattGal) [inline src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs]: So much wonderful null checking, except this line and 53.  Is this guaranteed to have entries?

Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs]: It...really should be. If not, I would say that's most likely a Kusto ingestion error. I will wrap it in a try-catch and add logging about that.

Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs]: Done.

File: src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs (hunk 1 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs]: Done.

File: src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs [hunk 3]
```diff
@@ -48,6 +48,35 @@ private async Task WaitForHelixJobAsync(string jobName)
                 var finishedCount = workItems.Count(wi => wi.State == "Finished");
                 if (waitingCount == 0 && runningCount == 0)
                 {
+                    try
+                    {
+                        var workItemSummary = await HelixApi.Aggregate.JobSummaryMethodAsync(new string[] { "job.source" }, 1, filtername: jobName);
```
Comment (akoeplinger) [inline src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs]: it probably doesn't matter here but the wait implementation from dotnet-ci used to group by `job.name` instead: https://github.com/dotnet/dotnet-ci/blob/793e9fe542a3597101e4e4aa833e8a7b50ce4dbe/vars/waitForHelixRuns.groovy#L111

File: src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs [hunk 4]
```diff
@@ -48,6 +48,35 @@ private async Task WaitForHelixJobAsync(string jobName)
...
+                        
+                        int? numWorkItemFailures;
+                        workItemSummary[0].Data.WorkItemStatus.TryGetValue("fail", out numWorkItemFailures);
+                        if ((numWorkItemFailures ?? 0) > 0)
+                        {
+                            Log.LogError($"Job {jobName} had {numWorkItemFailures} work item(s) fail.");
+                        }
+
+                        int? numTestFailures;
+                        workItemSummary[0].Data.Analysis[0].Status.TryGetValue("fail", out numTestFailures);
```
Comment (akoeplinger) [inline src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs]: there's a timeframe where the xunit results are not fully processed yet, I think you'll need to wait until that part is done or some failed tests might be missed here. See similar code in https://github.com/dotnet/dotnet-ci/blob/793e9fe542a3597101e4e4aa833e8a7b50ce4dbe/vars/waitForHelixRuns.groovy#L154-L158

Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/HelixWait.cs]: Yup, you're 100% right. I noticed this bug yesterday during testing and was going to make changes to fix it today. Thanks!

**Pull request closed without merging**